File: src/main/java/io/rsocket/cli/util/HeaderUtil.java
package io.rsocket.cli.util;

import com.google.common.base.Charsets;
import com.google.common.io.CharSource;
import com.google.common.io.Files;
import io.rsocket.cli.Publishers;
import io.rsocket.cli.UsageException;
import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static io.rsocket.cli.util.FileUtil.expectedFile;

// TODO handle duplicate header keys
public class HeaderUtil {
  public static Map<String, String> headerMap(List<String> headers) {
    Map<String, String> headerMap = new LinkedHashMap<>();

    if (headers != null) {
      for (String header : headers) {
        if (header.startsWith("@")) {
          headerMap.putAll(headerFileMap(header));
        } else {
          String[] parts = header.split(":", 2);
          // TODO: consider better strategy than simple trim
          String name = parts[0].trim();
          String value = stringValue(parts[1].trim());
          headerMap.put(name, value);
        }
      }
    }
    return headerMap;
  }

  private static Map<? extends String, ? extends String> headerFileMap(String input) {
    CharSource is = Files.asCharSource(inputFile(input), Charsets.UTF_8);
    return headerMap(Publishers.splitInLines(is).collectList().block());
  }

  public static String stringValue(String source) {
    if (source.startsWith("@")) {
      try {
        return Files.toString(inputFile(source), StandardCharsets.UTF_8);
      } catch (IOException e) {
        throw new UsageException(e.toString());
      }
    } else {
      return source;
    }
  }

  public static File inputFile(String path) {
    return expectedFile(path.substring(1));
  }
}

File: src/main/kotlin/io/rsocket/cli/util/HeaderUtil.kt
package io.rsocket.cli.util

import com.google.common.base.Charsets
import com.google.common.io.Files
import io.rsocket.cli.Publishers
import io.rsocket.cli.UsageException
import io.rsocket.cli.util.FileUtil.expectedFile
import java.io.File
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.util.*

// TODO handle duplicate header keys
object HeaderUtil {
  fun headerMap(headers: List<String>?): Map<String, String> {
    val headerMap = LinkedHashMap<String, String>()

    if (headers != null) {
      for (header in headers) {
        if (header.startsWith("@")) {
          headerMap.putAll(headerFileMap(header))
        } else {
          val parts = header.split(":".toRegex(), 2).toTypedArray()
          // TODO: consider better strategy than simple trim
          val name = parts[0].trim({ it <= ' ' })
          val value = stringValue(parts[1].trim({ it <= ' ' }))
          headerMap.put(name, value)
        }
      }
    }
    return headerMap
  }

  private fun headerFileMap(input: String): Map<out String, String> =
      headerMap(Publishers.splitInLines(Files.asCharSource(inputFile(input), Charsets.UTF_8)).collectList().block())

  fun stringValue(source: String): String = when {
    source.startsWith("@") -> try {
      Files.toString(inputFile(source), StandardCharsets.UTF_8)
    } catch (e: IOException) {
      throw UsageException(e.toString())
    }
    else -> source
  }

  fun inputFile(path: String): File = expectedFile(path.substring(1))
}
